### PR TITLE
add elasticsearch sink for heapster

### PR DESCRIPTION
**Pull request by huangyuqi**
This sink supports monitoring metrics and events.
To use the Elasticsearch sink add the following flag:

```
--sink=elasticsearch:<ES_SERVER_URL>[?<OPTIONS>]
```

Normally, elasticsearch cluster has multi nodes or a proxy, so nodes' list or proxy 
server need be configured for elasticsearch sink.
Thus, we can set `ES_SERVER_URL` to a dummy value, and provide nodes list or proxy server
in url's query string.
Besides, the following options can be set in query string:
- `index` - ES's index for metrics&events. Default: `heapster`
  
  if the ES cluster needs authentication, we should provide the username and secret.
- `esUserName`      - ES's user name
- `esUserSecret`    - ES's user secret

Like this:

```
--sink="elasticsearch:?nodes=0.0.0.0:9200&index=testMetric"

or

--sink="elasticsearch:?nodes=0.0.0.0:9200&index=testEvent"
```

### CODE REVIEW

**Comment by vishh** [comment]
ok to test

**Comment by vishh** [comment]
cc @kubernetes/heapster-maintainers

**Comment by vishh** [comment]
needs a rebase

**Comment by huangyuqi** [comment]
Hi, @vishh ,Is there something wrong about the CI? I got the error in checking process,

```
I1203 02:14:44.359848   15679 heapster_api_test.go:79] Failed to create ns: namespaces "heapster-e2e-tests" already exists
--- FAIL: TestHeapster (51.98s)
	assertions.go:154: 
                        
	Location:	heapster_api_test.go:784
		
	Error:		No error is expected but got namespaces "heapster-e2e-tests" already exists
```

**Comment by vishh** [comment]
Can you rebase to HEAD and try again?

**Comment by huangyuqi** [comment]
OK, I will try. Thanks @vishh

**Comment by huangyuqi** [comment]
Hi, @vishh, I have rebased, but the checks were still not successful.

**Comment by vishh** [comment]
cc @piosz. Is our test infra broken?

**Comment by piosz** [comment]
It seems to be broken. How can I get there to fix it?

**Comment by vishh** [comment]
ok to test

**Comment by huangyuqi** [comment]
Hi, @vishh , how can I resolve the issue about the above checks? Thanks.

**Comment by vishh** [comment]
@huangyuqi: @mwielgus might be able to assist you with that.

**Comment by huangyuqi** [comment]
thanks @vishh 
Hi, @mwielgus , how can I resolve the issue about the above checks? Thanks.

**Comment by vishh** [comment]
ok to test

**Comment by huangyuqi** [comment]
Hi, @vishh 
I have taken a rebase with master.would you please help to review ? Thanks.

File: Godeps/Godeps.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
 	"ImportPath": "k8s.io/heapster",
-	"GoVersion": "go1.5",
```
Comment (vishh) [inline Godeps/Godeps.json]: Dont change go version

File: Godeps/Godeps.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
```
Comment (vishh) [inline Godeps/Godeps.json]: Godep changes look a bit wonky. I'd suggest not updating any deps that this PR does not depend on.

**Comment by vishh** [comment]
@huangyuqi @mwielgus: This patch will be useful for shipping events to elasticsearch and view them via Kibana. Can we prioritize this?

**Comment by huangyuqi** [comment]
@vishh @mwielgus I will fix issues of the CI checks right now.
BTW, some guys are talking about where to get the events in issue https://github.com/kubernetes/kubernetes/issues/19637, would you please give some suggestions?

File: Godeps/Godeps.json (hunk 2 above)
Comment (huangyuqi) [inline Godeps/Godeps.json]: yes, maybe something wrong I have done during rebase.

**Comment by biswars** [comment]
Hi @huangyuqi , I just tried to test this commit with elasticsearch but getting connect error always.
Note: The elasticsearch service is reachable without user credentials using curl 172.31.0.71:9200

This is how I use it "./heapster -v=3 --source=kubernetes:https://kubernetes.default --sinelasticsearch:?nodes=172.31.0.71:9200"

The log shows the follows:
I0203 11:43:45.468969   22695 heapster.go:61] ./heapster -v=3 --source=kubernetes:https://kubernetes.default --sink=elasticsearch:?nodes=172.31.0.71:9200
I0203 11:43:45.469054   22695 heapster.go:62] Heapster version 0.19.1
E0203 11:43:45.469130   22695 heapster.go:122] failed to setup sinks - encountered following errors while setting up sources - open /var/run/secrets/kubernetes.io/serviceaccount/token: no such file or directory
I0203 11:43:45.469208   22695 driver.go:252] initializing elasticsearch sink with ES's nodes - [172.31.0.71:9200]
I0203 11:43:45.469239   22695 driver.go:194] attempting to setup elasticsearch sink
I0203 11:43:45.469373   22695 external.go:116] no timeseries data between 0001-01-01 00:00:00 +0000 UTC and 0001-01-01 00:00:00 +0000 UTC
E0203 11:43:50.470230   22695 client.go:58] Failed to initialize client "elasticsearch"- failed to connect any node of ES cluster
I0203 11:43:50.470332   22695 driver.go:194] attempting to setup elasticsearch sink
I0203 11:43:50.470475   22695 external.go:223] Updated sinks: [0xc20805ad00]
I0203 11:43:50.471495   22695 manager.go:162] starting to scrape data from sources start: 2016-02-03 11:43:45 +0000 UTC end: 2016-02-03 11:43:50 +0000 UTC
I0203 11:43:50.472490   22695 manager.go:175] completed scraping data from sources. Errors: []
I0203 11:43:50.473409   22695 heapster.go:72] Starting heapster on port 8082

I don't see any indices getting created in elaasticsearch and the line in the log above shows "Failed to initialize client "elasticsearch"- failed to connect any node of ES cluster"

**Comment by moserke** [comment]
Any progress on this?  Would be great to have an elasticsearch sink

**Comment by huangyuqi** [comment]
Hi, @biswars , I am on vacation this month , so sorry for delay. 
I will check this, because this PR is old, may be something changed with the base code.

**Comment by huangyuqi** [comment]
Hi, @moserke 
ES sink in this PR has been finished. Maybe some reviews is in process.
@mwielgus

**Comment by moserke** [comment]
@huangyuqi Have you looked at the modifications needed to grafana to produce the same dashboards in the grafana package with influx?  Wanted to see if you had anything there before digging in to it myself.

**Comment by titilambert** [comment]
@moserke I make a PR to be able to get events directly in InfluxBD: https://github.com/kubernetes/heapster/pull/1145
I also added a Grafana dashboard, called **events_withfields.json**, to get data from those events.

**Comment by piosz** [comment]
@huangyuqi  Is this ready for review?

**Comment by huangyuqi** [comment]
@piosz  It's ready, please give some review coments. Thanks so much

**Comment by piosz** [comment]
@mwielgus could you please take a look?

**Comment by piosz** [comment]
@biswars, @moserke, @titilambert if you are familiarized with elasticsearch could you please also take a look?

**Comment by titilambert** [comment]
@piosz the first thing I see, it seems there is code duplication (two driver.go and two driver_test.go files)
(functions: SaveDataIntoES, NewElasticSearchSink, ...)

**Comment by piosz** [comment]
@titilambert please comment in the code.

File: metrics/sinks/elasticsearch/driver.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+func (sink *elasticSearchSink) Name() string {
+	return "ElasticSearch Sink"
+}
+
+func (sink *elasticSearchSink) Stop() {
+	// nothing needs to be done.
+}
+
```
Comment (titilambert) [inline metrics/sinks/elasticsearch/driver.go]: NewElasticSearchSink is already defined in events/sinks

File: metrics/sinks/elasticsearch/driver.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+				MetricsValue: map[string]interface{}{
+					"value": metric.GetValue(),
+				},
+				MetricsTimestamp: dataBatch.Timestamp.UTC(),
+			}
+			sink.saveDataFunc(sink.esClient, sink.timeSeriesIndex, typeName, point)
+		}
+	}
+}
+
```
Comment (titilambert) [inline metrics/sinks/elasticsearch/driver.go]: SaveDataIntoES is already defined in events/sinks/elasticsearch

**Comment by huangyuqi** [comment]
Hi, @titilambert 
Now, in new version of heapster, events and metrics are divided into two different process. So, the elasticsearch sink has double. But the content putting to sink are different.

**Comment by piosz** [comment]
@huangyuqi we have a common directory for the common logic. For example for influxdb you can find it here https://github.com/kubernetes/heapster/tree/master/common/influxdb

**Comment by huangyuqi** [comment]
@piosz thanks, i will pick out the common part.

**Comment by titilambert** [comment]
@huangyugi you can check in #1144 I did this split for elasticsearch...

**Comment by piosz** [comment]
Any progress on this? It would be great to include it in Heapster 1.1/Kubernetes 1.3. The code freeze is in 2 weeks from now.

**Comment by huangyuqi** [comment]
Hi, @piosz  I have finished this ES sink, please give me some review comments. Thanks.

**Comment by huangyuqi** [comment]
I got an error in CI:

```
ok  	k8s.io/heapster/events/manager	5.528s
--- FAIL: TestAllExportsInTime (2.00s)
	<autogenerated>:31: 
                        
	Location:	manager_test.go:55
		
	Error:		Not equal: 3 (expected)
			        != 2 (actual)
		
W0512 06:35:16.299389   29333 manager.go:108] Failed to events data to sink: s2
W0512 06:35:19.309638   29333 manager.go:108] Failed to events data to sink: s2
W0512 06:35:22.310790   29333 manager.go:108] Failed to events data to sink: s2
W0512 06:35:22.310817   29333 manager.go:108] Failed to events data to sink: s1
W0512 06:35:25.311504   29333 manager.go:108] Failed to events data to sink: s2
W0512 06:35:25.311505   29333 manager.go:108] Failed to events data to sink: s1
FAIL
FAIL	k8s.io/heapster/events/sinks	15.047s

```

**Comment by piosz** [comment]
this is a know flake

**Comment by piosz** [comment]
@k8s-bot test this

**Comment by huangyuqi** [comment]
@piosz , some issues still exist in CI

```
[0;33mWaiting for 2 ready nodes. 0 ready nodes, 0 registered. Retrying.[0m
[0;33mWaiting for 2 ready nodes. 0 ready nodes, 0 registered. Retrying.[0m
[0;31mDetected 0 ready nodes, found 0 nodes out of expected 2. Your cluster may not be working.[0m
I0512 07:20:20.354669   22591 framework.go:285] Retry validation after 10ns seconds.
I0512 07:20:30.354878   22591 framework.go:212] validating existing cluster
I0512 07:20:30.354934   22591 framework.go:133] about to run &{/tmp/heapster_test/1.0.7/kubernetes/cluster/validate-cluster.sh [/tmp/heapster_test/1.0.7/kubernetes/cluster/validate-cluster.sh] []  <nil> <nil> <nil> [] <nil> <nil> <nil> <nil> false [] [] [] [] <nil>}
panic: test timed out after 30m0s

goroutine 35 [running]:
testing.startAlarm.func1()
	/usr/local/go/src/testing/testing.go:703 +0x132
created by time.goFunc
	/usr/local/go/src/time/sleep.go:129 +0x3a
```

**Comment by piosz** [comment]
@k8s-bot test this

**Comment by piosz** [comment]
@huangyuqi I fixed it yesterday #1159

**Comment by huangyuqi** [comment]
thanks so much @piosz

**Comment by ixdy** [comment]
@k8s-bot test this please

**Comment by mwielgus** [comment]
LGTM

**Comment by huangyuqi** [comment]
thanks so much @mwielgus

**Comment by mwielgus** [comment]
Thank you too :)

**Pull request merged**